cheatsheets/File_Upload_Cheat_Sheet.md

```diff
@@ -0,0 +1,163 @@
+
+- Double extensions, _e.g._ `.jpg.php`, where it circumvents easily the regex `\.jpg`
+- Null bytes, _e.g._ `.php%00.jpg`, where `.jpg` gets truncated and `.php` becomes the new extension
+- Generic bad regex that isn't properly tested and well reviewed. Refrain from building your own logic unless you have enough knowledge on this topic.
+
+Refer to the [Input Validation CS](Input_Validation_Cheat_Sheet.md) to properly parse and process the extension.
+
+#### Whitelist Extensions
+
+Ensure the usage of *business-critical* extensions only, without allowing any type of *non-required* extensions. If the system requires:
+
+- image upload, allow one type that is agreed upon to fit the business requirement;
+- cv upload, allow `docx` and `pdf` extensions.
+
+Based on the needs of the application, ensure the **least harmful** and the **lowest risk** file types to be used.
+
+#### Blacklist Extensions
+
+> Blacklisting extensions is a bad idea and is very dangerous. Don't do it unless you have no other choice!
+
+In order to perform this validation, specifying and identifying which patterns that could should be rejected are used in order to protect the service.
+
+### Content-Type Validation
+
+Other than defining the extension of the uploaded file, its MIME-type can be checked for a quick protection against simple file upload attacks.
+
+This can be done preferrably in a whitelist approach; otherwise, this can be done in a blacklist approach.
+
+> This should not be used on its own, as bypassing it is pretty common and easy.
+
+### Magic Number Validation
+
+In conjunction with [content-type validation](#content-type-validation), validating the file's magic numbers can be checked and verified against the expected file that should be received.
+
+> This should not be used on its own, as bypassing it is pretty common and easy.
+
+### File Name Sanitization
+
+Multiple threats relate to the file-name, thus, it is essential to ensure that the file-name is stored in a secure fashion.
+
+File-names can endager the system in multiple ways, either by using non acceptable characters, or by using special and restricted filenames, which can be found under the following [MSDN guide](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file?redirectedfrom=MSDN#naming-conventions). For a wider overview on different filesystems and how they treat files, refer to [Wikipedia's Filename page](https://en.wikipedia.org/wiki/Filename).
+
+In order to avoid the above mentioned threat, creating a **random string** as a file-name, _e.g_ generating a UUID/GUID, is essential. If the file-name is required by the business needs, proper input validation should be done for client-side (_e.g._ active content) and back-end side (_e.g._ special files overwrite or creation) attack vectors. File-name length limits should be taken into consideration based on the system storing the files, as each system has its own file name length limit.
```

Comment:
`such as` would read better than `e.g`

Not clear what an `active content` attack is in a filename. Do you mean things like XSS?

Some more practical guidance could be given here, such as:

- Implement a maximum length
- Restrict characters to alphanumeric, spaces, hyphens and periods (for example)
- Otherwise blacklist dangerous characters (non-printable, slashes, colons, etc) and patterns (two periods)